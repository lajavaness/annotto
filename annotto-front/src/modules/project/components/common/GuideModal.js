import { Col, Row, Switch, Typography } from 'antd'
import { EditOutlined, ExportOutlined } from '@ant-design/icons'
import { isEmpty } from 'lodash'
import { navigate } from '@reach/router'
import { useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import PropTypes from 'prop-types'
import React, { useCallback, useMemo, useState } from 'react'
import ReactMarkdown from 'react-markdown'

import WarningMessage from 'shared/components/warning/WarningMessage'

import useProjectPrivileges from 'shared/hooks/useProjectPrivileges'

import { selectProjectId } from 'modules/project/selectors/projectSelectors'

import { EDIT_PROJECT } from 'shared/enums/privilegesTypes'

import { groupTasksByCategories } from 'shared/utils/tasksUtils'

import 'github-markdown-css'

import * as Styled from './__styles__/GuideModal.styles'

const GuideModal = ({ guidelines, tasks, entitiesRelationsGroup }) => {
	const [isAnnotationGuide, setIsAnnotationGuide] = useState(false)

	const projectId = useSelector(selectProjectId)

	const { t } = useTranslation('project')

	const hasPrivilege = useProjectPrivileges()

	const _onModeChange = (checked) => setIsAnnotationGuide(checked)

	const _onEditClick = useCallback(
		() => navigate(`/edit_project/${projectId}/${isAnnotationGuide ? 'guide' : 'labeling'}`, { replace: true }),
		[isAnnotationGuide, projectId]
	)

	const resolvedTasksByCategories = useMemo(() => groupTasksByCategories(tasks), [tasks])

	const resolveSections = (key, name, children) => (
		<Styled.Group key={key}>
			<Typography.Title level={4}>{name}</Typography.Title>
			{!isEmpty(children) &&
				children.map(({ label, hotkey, description }, index) => (
					<Styled.Task key={`section_child_${label}_${index}`}>
						<Typography.Title level={5}>
							{label} {hotkey && `(${hotkey})`}
						</Typography.Title>
						<Styled.Description>{description}</Styled.Description>
					</Styled.Task>
				))}
		</Styled.Group>
	)

	return (
		<Styled.Root>
			<Styled.Header>
				<Row gutter={{ xs: 8, sm: 10, md: 20 }}>
					<Col>
						<Styled.Title $isActive={isAnnotationGuide}>{t('project:guide.glossary')}</Styled.Title>
					</Col>
					<Col>
						<Switch checked={isAnnotationGuide} onChange={_onModeChange} />
					</Col>
					<Col>
						<Styled.Title $isActive={!isAnnotationGuide}>{t('project:guide.guidelines')}</Styled.Title>
					</Col>
					<Col>
						<Styled.Divider type="vertical" />
					</Col>
					<Col>
						{hasPrivilege(EDIT_PROJECT) && (
							<Styled.Button type="link" icon={<EditOutlined />} onClick={_onEditClick}>
								{t('project:guide.edit')}
							</Styled.Button>
						)}
					</Col>
					<Styled.LastCol>
						<Styled.Button type="link" icon={<ExportOutlined />} />
						<Styled.Divider type="vertical" />
					</Styled.LastCol>
				</Row>
			</Styled.Header>
			{isAnnotationGuide ? (
				<Styled.MarkdownContainer className="markdown-body">
					{!isEmpty(guidelines) ? (
						<ReactMarkdown source={guidelines} />
					) : (
						<WarningMessage message={t('project:guide.emptyGuidelines')} />
					)}
				</Styled.MarkdownContainer>
			) : (
				<Styled.GlossaryContainer>
					{resolvedTasksByCategories.map(({ category, children }, index) =>
						resolveSections(`task_${category}_${index}`, category, children)
					)}
					{!isEmpty(entitiesRelationsGroup) &&
						entitiesRelationsGroup.map(({ name, values }, index) =>
							resolveSections(`entitiesRelation_${name}_${index}`, name, values)
						)}
				</Styled.GlossaryContainer>
			)}
		</Styled.Root>
	)
}

export default GuideModal

const EntitiesRelationShape = PropTypes.shape({
	/** The label of the entitiesRelation. Defines the text to displayed to define the entitiesRelation. */
	label: PropTypes.string.isRequired,
	/** A keyboard shortcut that is bound when the list is displayed to toggle
	 * an annotation. */
	hotkey: PropTypes.hotkey,
	/** The text displayed below the label and used to describe it. */
	description: PropTypes.string,
})

const TaskShape = PropTypes.shape({
	/** The label of the task. Defines the text to displayed to define the task. */
	label: PropTypes.string.isRequired,
	/** A keyboard shortcut that is bound when the list is displayed to toggle an annotation. */
	hotkey: PropTypes.hotkey,
	/** The text displayed below the label and used to describe it. */
	description: PropTypes.string,
	/** Used to group the different task and to display the name this group. */
	category: PropTypes.string,
})

GuideModal.propTypes = {
	/** Defines the source of markdown. */
	guidelines: PropTypes.string,
	/** The list of task available in the project. Used to displayed the label,
	 * the hotkey and the description of each task. */
	tasks: PropTypes.arrayOf(TaskShape),
	/** The list of entitiesRelationsGroup available in the project. Used to displayed the label,
	 * the hotkey and the description of each entitiesRelations. */
	entitiesRelationsGroup: PropTypes.arrayOf(
		PropTypes.shape({
			/** Used to group the different relations and to display the name this group. */
			name: PropTypes.string,
			values: PropTypes.arrayOf(EntitiesRelationShape),
		})
	),
}

GuideModal.defaultProps = {
	guidelines: null,
	tasks: null,
	entitiesRelationsGroup: null,
}
