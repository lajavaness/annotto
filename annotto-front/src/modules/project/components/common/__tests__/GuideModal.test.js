import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import GuideModal from 'modules/project/components/common/GuideModal'

const initialState = {}
const defaultProps = {}

const getStore = (state) => createStore({ initialState: state })

const getInstance = (props = {}, state = initialState) => (
	<Provider store={getStore(state)}>
		<ThemeProvider theme={theme}>
			<GuideModal {...defaultProps} {...props} />
		</ThemeProvider>
	</Provider>
)

describe('GuideModal', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
