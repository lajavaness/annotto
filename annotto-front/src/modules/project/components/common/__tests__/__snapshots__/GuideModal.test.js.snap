// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuideModal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-kmATbt"
  >
    <div
      class="sc-cvJIAo BmuxP"
    >
      <div
        class="ant-row"
        style="margin-left: -10px; margin-right: -10px;"
      >
        <div
          class="ant-col"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <span
            class="sc-gIRjGK fRehlE"
          >
            project:guide.glossary
          </span>
        </div>
        <div
          class="ant-col"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <button
            aria-checked="false"
            class="ant-switch"
            role="switch"
            type="button"
          >
            <div
              class="ant-switch-handle"
            />
            <span
              class="ant-switch-inner"
            />
          </button>
        </div>
        <div
          class="ant-col"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <span
            class="sc-gIRjGK hleETk"
          >
            project:guide.guidelines
          </span>
        </div>
        <div
          class="ant-col"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <div
            class="ant-divider ant-divider-vertical sc-dkmKIT XjYIs"
            role="separator"
          />
        </div>
        <div
          class="ant-col"
          style="padding-left: 10px; padding-right: 10px;"
        />
        <div
          class="ant-col sc-iWFTwQ eslkyd"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <button
            class="ant-btn ant-btn-link ant-btn-icon-only sc-fybuyZ gVgChl"
            type="button"
          >
            <span
              aria-label="export"
              class="anticon anticon-export"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="export"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M888.3 757.4h-53.8c-4.2 0-7.7 3.5-7.7 7.7v61.8H197.1V197.1h629.8v61.8c0 4.2 3.5 7.7 7.7 7.7h53.8c4.2 0 7.7-3.4 7.7-7.7V158.7c0-17-13.7-30.7-30.7-30.7H158.7c-17 0-30.7 13.7-30.7 30.7v706.6c0 17 13.7 30.7 30.7 30.7h706.6c17 0 30.7-13.7 30.7-30.7V765.1c0-4.3-3.5-7.7-7.7-7.7zm18.6-251.7L765 393.7c-5.3-4.2-13-.4-13 6.3v76H438c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h314v76c0 6.7 7.8 10.5 13 6.3l141.9-112a8 8 0 000-12.6z"
                />
              </svg>
            </span>
          </button>
          <div
            class="ant-divider ant-divider-vertical sc-dkmKIT XjYIs"
            role="separator"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-bUrLeq fquejC"
    />
  </div>
</DocumentFragment>
`;
